hw/drivers/lp5523/src/lp5523.c

```diff
@@ -0,0 +1,1144 @@
+}
+
+/**
+ * Writes to a register with an engine-based address offset,
+ * useful to write to:
+ * LP5523_ENGINE_PC
+ * LP5523_ENG_PROG_START_ADDR
+ * LP5523_MASTER_FADER
+ *
+ * @param The sensor interface
+ * @param The register address to write to (the MSB register)
+ * @param Value to write
+ *
+ * @return 0 on success, non-zero error on failure.
+ */
+int
+lp5523_set_engine_reg(struct sensor_itf *itf, enum lp5523_engine_registers addr,
+    uint8_t engine, uint8_t value)
+{
+    --engine;
+    return (engine > 2) ? -1 : lp5523_set_reg(itf, addr + engine, value);
+}
+
+/**
+ * Reads from a register with an engine-based address offset,
+ * useful to read from:
+ * LP5523_ENGINE_PC
+ * LP5523_ENG_PROG_START_ADDR
+ * LP5523_MASTER_FADER
+ *
+ * @param The sensor interface
+ * @param The register address to read from (the MSB register)
+ * @param Pointer to where the register values should be written
+ *
+ * @return 0 on success, non-zero error on failure.
+ */
+int
+lp5523_get_engine_reg(struct sensor_itf *itf, enum lp5523_engine_registers addr,
+    uint8_t engine, uint8_t *value)
+{
+    --engine;
+    return (engine > 2) ? -1 : lp5523_get_reg(itf, addr + engine, value);
+}
+
+/**
+ * Sets the CHIP_EN bit in the ENABLE register
+ *
+ * @param The sensor interface
+ * @param Non 0 to enable the device, 0 to disable it
+ *
+ * @return 0 on success, non-zero error on failure.
+ */
+int
+lp5523_set_enable(struct sensor_itf *itf, uint8_t enable)
+{
+    int rc;
+
+    rc = lp5523_set_val(itf, LP5523_ENABLE, LP5523_CHIP_EN, enable);
+
+    if ((rc == 0) && (enable)) {
+        lp5523_wait(1);
+    }
+    return rc;
+}
+
+/**
+ * Sets the ENGINEX_MODE and ENGINEX_EXEC bits in the ENGINE CNTRLX registers
+ *
+ * @param The sensor interface struct
+ * @param Control register to update: LP5523_ENGINE_CNTRL1 or
+ *  LP5523_ENGINE_CNTRL2
+ * @param Engines to update, a combination of LP5523_ENGINE3, LP5523_ENGINE2
+ *  or LP5523_ENGINE1
+ * @param Values to write, the lower 6 bits are relavant, 2 per engine
+ *
+ * @return 0 on success, non-zero error on failure.
+ */
+int
+lp5523_set_engine_control(struct sensor_itf *itf, enum lp5523_registers addr,
+    uint8_t engines, uint8_t values)
+{
+    uint8_t reg;
+    int rc;
+
+    if (((engines & LP5523_ENGINE1) != LP5523_ENGINE1) &&
+        ((engines & LP5523_ENGINE2) != LP5523_ENGINE2) &&
+        ((engines & LP5523_ENGINE3) != LP5523_ENGINE3)) {
+        return -1;
+    }
+
+    rc = lp5523_get_reg(itf, addr, &reg);
+    if (rc != 0) {
+        return rc;
+    }
+
+    reg &= ~engines;
+    reg |= engines & values;
+
+    return lp5523_set_reg(itf, addr, reg);
+}
+
+/**
+ * Sets the MAPPING bits in the relevant DX CONTROL register
+ *
+ * @param The sensor interface struct
+ * @param Output number 1-9
+ * @param Mapping 0-3
+ *
+ * @return 0 on success, non-zero error on failure.
+ */
+int
+lp5523_set_op_mapping(struct sensor_itf *itf, uint8_t op, uint8_t mapping)
+{
+    --op;
```

Comment:
What if op is 0 ? Is that possible ? There are multiple places in the code where op is decremented before getting used. It would be nice to cap it at 0 or add #defines for specific values OP would have, it's because these APIs seem to be public.

Reply:
I am not sure what you mean by "cap it at 0"? I can make them 0-relative if you like, or add defines as you suggested (or both).

Reply:
Adding defines sounds like the right approach. By cap it at 0, I meant make 0 look like an invalid value explicitly.